### ISSUE TITLE
audio rulers

### ISSUE THREAD

**Issue by ahicks92** [Factorio-Access/FactorioAccess#161]
## **1. Is your feature request related to a problem? Please describe.**

Sort of: if you have to zigzag, getting back to the original line is slow and this gets worse the further away you get.  If using build lock, undercounting is a problem and overcounting can weirdly be worse if you sideload another belt.

For an example look at my attached save.  There's the rail line on the east and the pumpjack on the west.  To get across around all those cliffs and whatnot requires maintaining the horizontal line across all of those cliffs even when you've got to turn and also use underground belts.

## **2. Describe the solution you'd like**

Something I've talked about on Discord a few times is "cursor stops".  We figured out that calling them audio rulers is way better and now that I have an example save for us to discuss somewhat it seemed like the right time.  There's a lot audio rulers could do, but this is what I think the basic flow should look like:

- You set an audio ruler like a bookmark, for now.  I think it'd actually be beyond useful to be able to put them in blueprints so it is maybe worth literally using an entity (consider refineries or factory modules where the below lines up with their inputs and outputs, and sharing blueprints among blind people).
- They can be optionally labeled and we'll announce which one in the case of lining up with it (see below).
- Audio rulers have two modes.  The first one is the one I want but the other one is intriguing and I can think of lots of uses like long train tracks, so I'll include it.
- The modes:
  - The easy one I want. Place one, set a range which we can just do by like "click for 10 until 100, 100 till 1000" or something.  Then it projects lines in the 4 cardinal directions out to that range like lasers or something to speak visually, all 1 tile wide.  That works for "coming in" from far away as well as more local alignment issues.
  - Alternatively click a start and an end aligned pair of points, and get only the one line.
- When you approach the line established by the ruler you get a sound.  Say within 5 tiles, preferably one that changes (e.g. just pitch the same thing up or down, we can do that with variants of the same sound file and can use ffmpeg to automate the problem).  For this volume may also work since the location of the sound isn't important.
- When you actually hit it you get a different sound.
- When you move along it with your cursor "inside", maybe we use a different sound or do nothing; if we do add a sound we probably want a different subtle one for leaving the line.
- We'll have to figure out pleasant sounds for this that aren't super annoying and alarming; you're supposed to put your cursor in that region.
- When you try to cross the line with cursor skipping, it doesn't cross over it. You have to do that by hand. That's how you take a local measurement if you really want the number.  Skip toward, get range, skip back or return.
- When build lock is on, build lock turns off at the ruler.  You don't get to build lock across them, but we should allow build loocking along the line itself.
  - This is because rulers are your "I must make a decision" line.  Otherwise why are you using them for the most part?  They solve other problems but not making mistakes is a big one.

So, how you can use it, and why I think it justifies the complexity:

- Place a cursor stop on the end of a transport belt. Turn build lock on and spam.  Hell maybe let's add a "to the ruler" build mode.  When done you're aligned with the other belt.  Solves building corners without a dedicated tool.  Also works for pipes.
- For cases like zigzagging near a line, you always know if you're closer than 5 tiles or whatever number we decide to do sounds to.
- When building vertical or horizontal things that have a vertical or horizontal thing they're related to but far apart, you can set up rulers and find out the alignment relative to you without having to go check, which is quite a few keys:
  - This is confusing so to elaborate with an actual workflow...
  - You build a "tall" oil refinery.
  - Plastic and other things go out to engines and electric engines, then later batteries etc.  Pretend it's a few horizontal lines of assemblers way off "over there".
  - You want to align the outputs with the inputs.
  - So, to do it you place a ruler on every east/west output of the assemblers.
  - Then you build from south to north in the refinery.
  - As you do, you get good notifications of where you are.
  - And finally, you can now cursor skip the outputs just by finding a long segment of pipe or dirt and skipping along it.

There's a lot of other places I can see it being used.  One other thing it does is lets your cursor move away from the line to go check on something without having to go nuts finding it again.  Another example is that you can place a ruler 2 tiles from a machine to know where to bring the belt in to be lined up for inserters.

There may be cases for other tools that specifically solve a specific problem, but this would solve most of those.  Indeed while it might make sense to control the width of the ruler as well as the length, you could even get that just by using two of them side to side.

## **3. Describe alternatives you've considered, if any, perhaps including previous discussions**

Basically everything else the mod does.  This is one option for taking a lot of smaller isolated pieces and pulling them together into a workflow that scales in a way which doesn't require counting off accurately and leads to mistakes like mixing belts and having to go fix the 200-unit lane because it's now holding 3 items, etc..:

- Basically no more counting.
- Cursor skipping doesn't require making and remembering long lines of stuff to skip along.
- Poor man's corner tool but good enough that I'd probably just use it over such a tool.

That's just not something we've really started on.

## **4. Additional context**

See my save.

There's a lot of room for tools like this but this is the immediately useful one.  As one other future direction, projecting grids of these lines has other good uses, for example if you can do 3x3 then you now have the ability to place assemblers and lanes between them and that literally would mean having two blueprints to alternate between and "fill out the tiles".

For a use case for 3x3 with a 1-tile gab between them, see my lab array (it's flawed in other ways but...).

100% not saying we should do this, just offering what I consider obvious future directions for such things. We should see how much people like the first version.

Not sure how to teach this.  There's probably lots of places in the tutorial we could fit it in.  Early ones at that.  Depends on if people find this to be an unnecessary digression or something super useful, but that's chicken-and-egg because first they need to know how...

[ruler.zip](https://github.com/user-attachments/files/15507673/ruler.zip)

**Comment by ahicks92** [comment]
So here's another use case, and in particular a use case for two instead of just one.  If we had a way to specifically skip to the next ruler in a given direction, you can line a bus with them.  The typical bus doesn't work well with other kinds of cursor skipping because it's belts with gaps, so my 70-something tile wide bus is like 20 skips.  That is, another way to view this is bookmarks for lines.  Like how bowling alleys have gutters to use a somewhat visual metaphor.

I could see up to 4 getting involved: west side/basic plates/fluids/right side.  Makes a "category selector" for lack of better phrasing.

**Comment by LevFendi** [comment]
So this week i made a first prototype by converting the single cursor bookmark into a basic audio ruler.

## features so far

It uses the game's own programmable speaker instrument sounds, with different instruments for cardinals versus diagonals, and different pitch for full aligment versus 1-tile-off-alignment.

You turn it on by setting the bookmark and vice versa. It has no range limit.

## Improvements quick list
1. Cursor skipping should be interrupted when intersecting with the outer edge of a ruler. Same for build lock.
1. The mod should support at least two bookmarks to enable more use cases
1. You should be able to control the ruler length for each bookmark separately, including zero for turning it off. Maybe just allow typing it in.
1. We can use better sounds, like maybe the speaker piano group.

## bookmarks versus fast travel points
We might want to name bookmarks too but then we are already blurring the boundaries between bookmarks and travel points. Perhaps we should combine the features. The remaining issue after that would be adding travel categories so that the list is shorter.

**Comment by ahicks92** [comment]
Yeah, let's just move discussion here.  You captured Discord above.

I think we should move the sounds to be "thematically correct".  The rest of the mod has done a good job at that.  The current option isn't really so great as has already been discussed, but the other thing is that this would be the only place in the mod which isn't matching theme, and it's a bit jarring.  Weirdly the diagonals sort of make sense.  Also Factorio is weird about sound too, in that even things like constant hums aren't used when they'd make sense.  I have actually weirdly enough heard a real-life steam turbine and after that the steam engines are quite, quite amusing.  If I had to describe the Factorio "theme", I'd say quiet synth percussive.  I will look into this.  If the programmable speaker isn't sufficient I believe I may have some public domain sounds which would be appropriate around.

I don't understand the use for diagonals which is code for let's drop them but you probably had a use case in mind.  If you imagine how two vertical rulers would intersect you can see that what really begins to happen is the formation of a square grid with squares rotated to 45 degrees on top of a grid with squares in the cardinal direction.  Add horizontal ones, and it starts to become a grid of *triangles*.  All of non-uniform size.  Since the only thing that can move at an angle is trains, I'm not sure that these are useful in any way.  But if there is a use case do share, it may very well be useful to even myself.

**Comment by ahicks92** [comment]
I verified that pan issues are from programmable speaker's sound base itself.

Few things in there that might be more fitting.  Few things in there that might also be workable for prototypes of graphing stuff.  But sadly few is doing much heavy lifting, in my opinion.

**Comment by LevFendi** [comment]
As another short term improvement before releasing our current version, I propose that we modify `CONTROL + SHIFT + B` to toggle the ruler functions on or off rather than to clear the bookmark (which happens anyway when you set a new one).

**Comment by LevFendi** [comment]
In PR #223 for Update 0.13.0 we have released a prototype version of a single audio ruler. It needs to be interfaced with cursor skipping and build lock mode but it works.
